addons/partner_autocomplete/models/res_company.py

```diff
@@ -1,14 +1,128 @@
+        if company_data.get('error'):
+            return {'error': company_data.get('error_message', _('Unkown reason'))}
+
+        additional_data = company_data.pop('additional_info', False)
+
+        if company_data.get('logo'):
+            # Erase image_1920 even if something is in it. Indeed as partner_autocomplete is probably installed as a
+            # core app (mail -> iap -> partner_autocomplete auto install chain) it is unlikely that people already
+            # updated their company logo.
+            try:
+                company_data['image_1920'] = base64.b64encode(
+                    requests.get(company_data['logo'], timeout=COMPANY_AC_TIMEOUT).content
+                )
+            except Exception as e:
+                _logger.warning('Download of image for new company %r failed, error %r' % (self.name, e))
+
+        # Keep only truthy values that are not already set on the target partner
+        company_data = {field: value for field, value in company_data.items()
+                        if field in self.partner_id._fields and field != 'image_1920' and value and not self.partner_id[field]}
+
+        company_data = self._enrich_extract_ids(company_data, ['state_id', 'country_id'])
+        if company_data.get('child_ids'):
+            child_ids = []
+            for child_data in company_data['child_ids']:
+                child_ids.append(self._enrich_extract_ids(child_data, ['state_id', 'country_id']))
+            company_data['child_ids'] = child_ids
+
+        # handle o2m values. Returned in the form: E.g. {'child_ids': [{'name': 'Value 1'}, {'name': 'Value 2'}]}
+        for o2m_field, create_values in company_data.items():
+            if isinstance(create_values, list):
+                commands = [(
+                    0, 0, create_value
+                ) for create_value in create_values if isinstance(create_value, dict)]
+
+                if commands:
+                    company_data[o2m_field] = commands
+                else:
+                    del company_data[o2m_field]
+
+        self.partner_id.write(company_data)
+        if additional_data:
+            template_values = json.loads(additional_data)
+            template_values['flavor_text'] = _("Company auto-completed by Odoo Partner Autocomplete Service")
+            self.partner_id.message_post_with_view(
+                'iap_mail.enrich_company',
+                values=template_values,
+                subtype_id=self.env.ref('mail.mt_note').id,
+            )
+        return {}
+
+    def _enrich_extract_ids(self, data, fields):
+        """ Extract country and state ids from data (because of res.partner._format_data_company)
+         This method is called during _enrich company process and is used twice:
+         For company_data itself and for child_ids """
+        for field in fields:
+            if isinstance(data.get(field), dict):
+                data[field] = data.get(field).get('id')
```

Comment:
`data.get` not necessary here, we already know the key is present. Also can it really be missing an `id` key?

Reply:
Cleaned while being at it